### PR TITLE
BSM SDK page

### LINKED ISSUE #2954: Document BSM JS SDK
### Description

The BSM feature relies on a JavaScript SDK. We should have a reference or guide that describes how this BSM JS SDK works, with examples.

Maybe this would look similar to [event filter expressions](https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-filter/filters/#build-event-filter-expressions-with-javascript-execution-functions). List of available functions + some examples for different use cases.

### Category

This is an issue with:

- [ ] errata (error in doc content)
- [x] new doc (requires a completely new doc)
- [ ] update (add missing or refresh existing content)
- [ ] improvement (improve docs content or format)

### Affected pages
This belongs in the same section as the rule template and components references (I think a sub-category within Operations, similar to how we've handled secrets).

### Expected behavior
We have a dedicated page in the docs to describe the BSM JavaScript SDK, with examples.

### Current behavior
We have not documented the BSM JavaScript SDK.

### Corrections needed
Add a BSM JS SDK doc.

### Additional context
https://github.com/sensu/sensu-enterprise-go/pull/1364
[BSM proposal](https://docs.google.com/document/d/1eMwAt2oHLh-lrWev6YicyiuzKvFF-dx7FWXV5DWem9Y/edit#)

### PR DESCRIPTION
**Pull request by hillaryfraley**
## Description
Adds a dedicated page for the BSM JavaScript SDK. Page is based on https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-filter/sensu-query-expressions/.

## Motivation and Context
Closes https://github.com/sensu/sensu-docs/issues/2954

## Review Instructions
This page is currently mainly an outline with information I pulled from https://github.com/sensu/sensu-enterprise-go/pull/1364. It may have some gaps like missing examples and making sure my assumptions based on the SQE reference are correct for the BSM SDK. See comments throughout draft.

### CODE REVIEW

File: content/sensu-go/6.3/operations/business-service-monitoring-sdk.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    parent: operations
+---
+
+Sensu's business service monitoring (BSM) feature uses a dedicated SDK of JavaScript-based expressions that provide additional functionality.
+
+BSM SDK expressions are defined in [rule templates][3], so they act in the context of determining whether aggregate data derived from a component’s selection of Sensu Go events should trigger a rule-based event.
+BSM SDK expressions always receive a single event and some information about that event, like `event.timestamp` or `event.check.interval`.
+
+BSM SDK expressions always return either `true` or `false`.
+They are evaluated by the [Otto JavaScript VM][1] as JavaScript programs.
```
Comment (hillaryfraley) [inline content/sensu-go/6.3/operations/business-service-monitoring-sdk.md]: These last three sentences are guesses based on the SQE reference (https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-filter/sensu-query-expressions/)

File: content/sensu-go/6.3/operations/business-service-monitoring-sdk.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+## Syntax quick reference
```
Comment (hillaryfraley) [inline content/sensu-go/6.3/operations/business-service-monitoring-sdk.md]: This is also copied from the SQE reference. I don't know if it's accurate for the BSM SDK.

File: content/sensu-go/6.3/operations/business-service-monitoring-sdk.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+<td>Greater than or equal to</td>
+</tr>
+</tbody>
+</table>
+
+## Specification
+
+BSM SDK expressions are valid ECMAScript 5 (JavaScript) expressions that return either `true` or `false`.
+Other values are not allowed.
+If an expression returns a value besides `true` or `false`, an error is recorded in the [Sensu backend log][2] and the filter evaluates to `false`.
```
Comment (hillaryfraley) [inline content/sensu-go/6.3/operations/business-service-monitoring-sdk.md]: From the SQE reference -- not sure if this is correct for the BSM SDK.

File: content/sensu-go/6.3/operations/business-service-monitoring-sdk.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+The BSM SDK allows you to to express rules for the number or percentage of events with critical, warning, OK, and unknown statuses.
+You can also configure expressions to ignore silenced events.
+
+## Custom functions
```
Comment (hillaryfraley) [inline content/sensu-go/6.3/operations/business-service-monitoring-sdk.md]: These are pulled from https://github.com/sensu/sensu-enterprise-go/pull/1364. I don't know if they are still correct.

File: content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+This BSM SDK expression returns `true` if ...:
+
+{{< code go >}}
+...
+{{< /code >}}
+
+
+[1]: https://github.com/robertkrimen/otto
+[2]: ../../observe-schedule/backend/#event-logging
```
Comment (hillaryfraley) [inline content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md]:
```suggestion
[2]: ../backend/#event-logging
```

File: content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+The custom function `sensu.Percentage()` returns ...
+Sunday is `0`.
+
+For example, ...:
+
+{{< code go >}}
+...
+{{< /code >}}
+
+### sensu.NewEvent()
```
Comment (ccressent) [inline content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md]: We ended up removing that function from the SDK altogether!

File: content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+The custom function `sensu.NewEvent()` returns ...
+
+For example, ...:
+
+{{< code go >}}
+...
+{{< /code >}}
+
+### sensu.NewMetricEvent()
```
Comment (ccressent) [inline content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md]: We ended up removing that function from the SDK altogether!

File: content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+The custom function `sensu.NewMetricEvent()` returns ...
+
+For example, ...:
+
+{{< code go >}}
+...
+{{< /code >}}
+
```
Comment (hillaryfraley) [inline content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md]:
```suggestion
```

File: content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+This BSM SDK expression returns `true` if ...:
+
+{{< code go >}}
+...
+{{< /code >}}
+
+
+[1]: https://github.com/robertkrimen/otto
+[2]: ../backend/#event-logging
+[3]: ../filters/#build-event-filter-expressions-with-sensu-query-expressions
```
Comment (hillaryfraley) [inline content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md]:
```suggestion
[3]: ../rule-templates/
[4]: ../../../commercial/
```

File: content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+description: "Based on JavaScript expressions, Sensu's business service monitoring SDK provides additional functionality for Sensu rule templates that evaluate service components. Read the reference doc to learn about the business service monitoring SDK."
+weight: 75
+version: "6.3"
+product: "Sensu Go"
+platformContent: false 
+menu:
+  sensu-go-6.3:
+    parent: observe-schedule
+---
+
```
Comment (hillaryfraley) [inline content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md:13]:
```suggestion

**COMMERCIAL FEATURE**: Access business service monitoring (BSM) in the packaged Sensu Go distribution.
For more information, see [Get started with commercial features][4].

```

File: content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+The custom function `sensu.Percentage()` returns the percentage of events with the specified status.
+For example, to return the percentage of events with `critical` status:
+
+{{< code go >}}
+sensu.Percentage("critical")
+{{< /code >}}
+
+## Example BSM SDK expression
+
```
Comment (hillaryfraley) [inline content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md:118]: I added this example based on notes I found from [Simon's spike](https://groups.google.com/a/sensu.io/g/eng-discuss/c/ok8rexkqi0E/m/YbUjVAjwBQAJ?pli=1)

**Review by echlebek — COMMENTED** [review]
I'm not sure if they're documented, but we should include some breadcrumbs to the Javascript functions that allow retrieving events from the event store. These were previously used to implement some advanced filters.

**Comment by hillaryfraley** [comment]
@echlebek do you mean https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-filter/sensu-query-expressions/?

**Review by echlebek — APPROVED** [review]

File: content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+**COMMERCIAL FEATURE**: Access business service monitoring (BSM) in the packaged Sensu Go distribution.
+For more information, see [Get started with commercial features][4].
+
+Sensu's business service monitoring (BSM) feature uses a dedicated SDK of JavaScript-based expressions that provide additional functionality.
+Use the BSM SDK to create custom JavaScript expressions with complex logic.
+
+BSM SDK expressions are defined in [rule templates][3], so they act in the context of determining whether aggregate data derived from a service component’s selection of Sensu Go events should trigger a rule-based event.
+They always receive a single event and some information about that event, like `event.timestamp` or `event.check.interval`, and always return either `true` or `false`.
+
+BSM SDK expressions are evaluated by the [Otto JavaScript VM][1] as JavaScript programs.
```
Comment (hillaryfraley) [inline content/sensu-go/6.3/observability-pipeline/observe-schedule/business-service-monitoring-sdk.md:27]:
```suggestion
BSM SDK expressions are evaluated by the [Otto JavaScript VM][1] as JavaScript programs.

{{% notice note %}}
**NOTE**: [Sensu query expressions](../../observe-filter/sensu-query-expressions/) also provide JavaScript functions for using nested parameters and custom functions to retrieve events from the event store.
{{% /notice %}}
```

**Pull request merged**